Connect: Add config file

gzdunek opened the pull request:
It is a draft PR to collect your early feedback on the updated `ConfigService`. Do you see any problems in maintaining it in the future?

The thing I really like is that paths and values are fully typed, so you can't make a typo in a (deep!) property name.

ravicious reviewed (commented):
It looks good enough to solve the issue at hand (persisting the choice regarding telemetry).

In terms of long-term maintenance, I think this PR needs to contain more details on the motivations behind certain choices. I see the code that you wrote but I lack information on why we need support for arbitrary nesting or why Lodash `merge` was chosen over other approaches or what `isStored` is useful for.

If you've based the implementation on something, it'd be good to know what the thing was!

avatus commented on packages/teleterm/src/services/config/types.ts:
```diff
@@ -2,17 +2,38 @@ import { Platform } from '../../mainProcess/types';
+export type ConfigGetter = <Path extends FieldPath<Config>>(
+  path: Path
+) => ConfigValue<Path>;
+export type ConfigSetter = <Path extends FieldPath<Config>>(
```
```suggestion

export type ConfigSetter = <Path extends FieldPath<Config>>(
```

avatus commented:
Overall, I'm a fan. As we discussed earlier, by the time this PR gets in, it'd be nice to have documentation on how to reset, or completely remove any config options (in the case of an uninstall) updated in the README somewhere.

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,83 @@
+  private readonly defaultConfig: Config;
+  private userConfig: Partial<Config>;
+  private mergedConfig: Config;
+  private version = '1';
```
My idea was to only have a placeholder for a version to use in the future, but now, after rethinking it, I am in favor of removing this. We can always add it in the future.

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,85 @@
+  get: ConfigGetter = path => {
+    return {
+      value: lodashGet(this.mergedConfig, path),
+      metadata: { isStored: !!lodashGet(this.userConfig, path) },
```
`isStored` will be useful for deciding if want to show the dialog about usage metrics. 
`isStored === false` means that a user didn't answer, so w should show the dialog, as in the example:
```typescript
// only if we didn't ask
  if (!configService.get('usageMetrics.enabled').metadata.isStored) {
   // show modal 
    configService.set('usageMetrics.enabled', response);
  }
```

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,85 @@
+  get: ConfigGetter = path => {
+    return {
+      value: lodashGet(this.mergedConfig, path),
+      metadata: { isStored: !!lodashGet(this.userConfig, path) },
```
🤦‍♂️

I wanted to add tests after the _draft_ phase, I would have caught that!

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,85 @@
+
+  private readUserConfig(): Partial<Config> {
+    // TODO(gzdunek): validation
+    return cloneDeep(this.appConfigFileStorage.get('config') || {});
```
That was my plan, so we could keep `version` and `config` in independent keys. Now I'm less sure about it, we probably don't need additional level of nesting.

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,85 @@
-  update(newConfig: Partial<Config>): void {
-    this.config = merge(this.config, newConfig);
-  }
+  set: ConfigSetter = (path, value) => {
```
At the beginning I thought it will be easier to keep it nested, because our config structure is nested already, so no major changes will be needed. 
But now I see the drawbacks of the nested structure: 
- we need to use Lodash getters/setters, 
- we need to merge the configs, which is not that simple. 

In case of the flat structure we only need to answer a simple question: "do we have that property stored in a file?" If yes - use it, if no - take the default.  

So now I think we should go with the flat structure.

I'm only wondering how to combine it with the current config providers, I like that different parts of config are separated in different files. I see two options here:
- add a top level key to each entry (`keyboardShortcuts.tab-1`, `keyboardShortcuts.tab-2`), but it is a lot of repetition 
- flatten the config in `configService`, but it requires more work to keep it typesafe.

I'm trying the second option, I will see how it works.

ravicious replied:
I'd also check how those two projects approach this, it seems that the structure might be "semi-flat", that is the keys in the config are flag but the actual config used by the app is not, idk.

gzdunek replied:
I will check that, but it seems to me that we should only work on flat configs. 
Let's say I have a following config type:
```
appearance: {
  fonts: {
   sansSerif: string
  }
}
```

In the code you can then change the config by: `set('appearance.fonts.sansSerif', 'Arial')`  or `set('appearance.fonts', {sansSerif: 'Arial'})`. 
Only the first option should be allowed; we can't save the second one as a flat config - it mixes property name with the value.

By defining properties this way: 
```
appearance.fonts.sansSerif: string;
terminal.customEnv: object;
```
you can easily know what is a key and what is a value.

ravicious replied:
Right, but we still need to accept scalar values. The configs above seem to be fine with you providing a map for `terminal.integrated.env.osx` or an array for `coc.preferences.formatOnSaveFiletypes`. So essentially you can provide values only for the "leaves" of the config – but yeah that's something you'd need to check.

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,85 @@
+  }
+
+  private produceMergedConfig(): void {
+    this.mergedConfig = merge({}, this.defaultConfig, this.userConfig);
```
Yeah, I agree, `merge` is not a good direction, it doesn't work well for arrays too:
`_.merge({y: ['a', 'b', 'c', 'f', 'z']}, {y:['a', 'd', 'e', 'f']})` gives `y: ["a", "d", "e", "f", "z"]`

In our case we would rather want to replace the whole array.

gzdunek commented:
I decided to change the approach. I removed most of the previous code and replaced it with (I hope!) much simpler key-value config. I took a look at other apps, like JetBrains Fleet and it has quite a lot of config keys and they are all very simple. This also led me to think that our structure was unnecessarily nested (like `appearance.fonts.mono` - I think `monoFontFamily` is absolutely enough). I removed the config providers and moved all the default configs to the `createConfigService` in order to simplify the config and make it easier to validate. Generally, I think that this simpler structure should last for long, I don't expect our app to have even 50 config keys in the foreseeable future.

Validation is done with [zod](https://github.com/colinhacks/zod), a schema validation library. I was considering defining the schema using JSON schema and validate it with [ajv](https://github.com/ajv-validator/ajv), but after a discussion with Rafał I decided to go with `zod`. We don't really need the JSON schema (it is useful when you provide it to the editor - you can have autocomplete in JSON files); also `ajv` seems to be less secure than `zod` https://ajv.js.org/security.html#content-security-policy. Additionally, if we ever need to switch to JSON schema it will be very simple.

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,180 @@
-    appearance: appearanceConfigProvider,
-  };
+const appConfigSchema = z.object({
+  'keymap.tab-1': z.string(),
+  'keymap.tab-2': z.string(),
+  'keymap.tab-3': z.string(),
+  'keymap.tab-4': z.string(),
+  'keymap.tab-5': z.string(),
+  'keymap.tab-6': z.string(),
+  'keymap.tab-7': z.string(),
+  'keymap.tab-8': z.string(),
+  'keymap.tab-9': z.string(),
+  'keymap.tab-close': z.string(),
+  'keymap.tab-new': z.string(),
+  'keymap.tab-previous': z.string(),
+  'keymap.tab-next': z.string(),
+  'keymap.toggle-connections': z.string(),
+  'keymap.toggle-clusters': z.string(),
+  'keymap.toggle-identity': z.string(),
+  'keymap.open-quick-input': z.string(),
```
I'm wondering about converting them to camelCase (`keymap.tab1`) so they will match the last 3 keys.

gzdunek replied:
Done, although we should also rename the keyboard shortcuts that are used in the app, like `tab-1` (not config key `keymap.tab1`), but it will affect many files, so I would do it in another PR.

ravicious commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,180 @@
+  'keymap.toggle-identity': z.string(),
+  'keymap.open-quick-input': z.string(),
+  sansSerifFontFamily: z.string(),
+  monoFontFamily: z.string(),
```
TBH, I'm not sure if we want to make fonts and shortcuts configurable right now.

AFAIK we don't validate those strings in any way which means that it opens us to an attack vector where arbitrary CSS can be injected through the config file. And I'm not sure if there are any attack vectors related to keyboard shortcuts.

gzdunek replied:
Yeah, a better validation is needed. I disabled reading them from the file.

gzdunek commented on packages/teleterm/src/services/config/configStore.ts:
```diff
@@ -0,0 +1,65 @@
+  Shape = z.infer<Schema>
+>(schema: Schema, defaults: Shape, fileStorage: FileStorage) {
+  const parsed = parse(fileStorage.get());
+  updateConfigFile();
```
Yeah, you're absolutely right, done.

gzdunek commented on packages/teleterm/src/services/config/configStore.ts:
```diff
@@ -0,0 +1,65 @@
+    fileStorage.putKey(key as string, value);
+  }
+
+  function getParsingErrors(): ZodIssue[] | undefined {
```
`getParsingErrors` can be a bit confusing I would like to just use a field here, but I can't because of IPC.
Maybe renaming it to `readValidationErrors` makes it more clear? 

I also removed updating the file.

ravicious replied:
> `getParsingErrors` can be a bit confusing I would like to just use a field here, but I can't because of IPC. Maybe renaming it to `readValidationErrors` makes it more clear?

Ah, I forgot about IPC. As you said, Without IPC even just `validationErrors` would look fine but it looks weird with IPC.

---

What about `getStoredConfigErrors`? This would make it even more clear that `getStoredConfigErrors` returns errors on which `get`/`set` have no bearing.

For me the biggest issue about this piece of code remains to be the fact that we do some side effects in one function call but we inspect the errors from it in a totally different place instead of returning them from the function call that performs those side effects. But I realize that with IPC there's no way around it.

That's why it's important for me to make the name of it as clear as possible to bring back some of this causality which would be present if this was all done in a single process.

gzdunek replied:
`getStoredConfigErrors` sounds totally fine to me, I don't have any strong opinion on the name, so I am happy to change it.

gzdunek commented on packages/teleterm/src/services/config/configStore.ts:
```diff
@@ -0,0 +1,65 @@
+    fileStorage.putKey(key as string, value);
+  }
+
+  function getParsingErrors(): ZodIssue[] | undefined {
```
>I'm not sure if this is something we want to tackle right now – AFAIK we won't be "advertising" the config file to the users right now, we want to just have a place to save the telemetry preference, right?

Exactly, definitely we should address, but I would like not to spend more time on this rn.

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,180 @@
+  // Defaults could be provided in the schema using .default(), but currently
+  // there is no way to remove them during parsing (so returned data that is
+  // saved to the file would contain all the properties).
+  // Come back to it when this is resolved https://github.com/colinhacks/zod/issues/1593
```
Correct.

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,180 @@
-  update(newConfig: Partial<Config>): void {
-    this.config = merge(this.config, newConfig);
+export type KeyboardShortcutsConfig = Record<KeyboardShortcutType, string>;
+const getKeymap = (platform: Platform) => {
```
Done

gzdunek commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,180 @@
-      },
-      {}
-    ) as Config;
+function getFonts(platform: Platform) {
```
Done

ravicious commented on packages/teleterm/src/services/config/configStore.ts:
```diff
@@ -8,58 +8,70 @@ const logger = new Logger('ConfigStore');
-      fileStorage.put(parsed.data); // parsed.data does not contain unknown keys
+      return {
+        storedConfig,
+        configWithDefaults: parsed.data as Shape,
```
Something about having to use `as` here rubbed me the wrong way since zod claims to be a TypeScript-first library.

Turns out we can change the definition of Schema to this:

```typescript
export function createConfigStore<
  Schema extends z.ZodTypeAny,
```

and then we can drop the `as` here and by `reParsed`. From what I checked, we still get properly inferred types.

See https://github.com/colinhacks/zod/tree/v3.20.2#writing-generic-functions

gzdunek replied:
Thanks for the link, I was sure I missed some obvious explanation how to write functions like this :) 

Changed.

avatus replied:
just a note here, zod rules!

ravicious commented on packages/teleterm/src/services/config/configStore.ts:
```diff
@@ -8,58 +8,70 @@ const logger = new Logger('ConfigStore');
+        configWithDefaults: parsed.data as Shape,
+        errors: undefined,
+      };
     } else {
```
We can remove one level of nesting by getting rid of this `else` here.

ravicious commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,203 @@
+    'fonts.monoFamily': omitStoredConfigValue(
+      z.string().default(defaultFonts['mono'])
+    ),
+    'usageMetrics.enabled': z.boolean().default(false),
```
I think it might be a good idea to put `usageMetrics.enabled` at the top of the schema definition since it's the only special value.

ravicious commented on packages/teleterm/src/services/config/configService.ts:
```diff
@@ -1,39 +1,203 @@
+  const defaultFonts = getDefaultFonts(platform);
+
+  // Important: all keys except 'usageMetrics.enabled' are currently not
+  // configurable by the user, we need to better validate the input (especially
+  // fonts, where any CSS can be injected).
```
```suggestion
  // configurable by the user. Before we let the user configure them,
  // we need to set up some actual validation, so that for example
  // arbitrary CSS cannot be injected into the app through font settings.
  //
  // However, we want them to be in the config schema, so we included
  // them here but we do not read their value from the stored config.
```

ravicious reviewed (commented):
I added some suggestions but nothing major. The current implementation should be good enough for the issue at hand and is extensible enough if we decide to offer an actual settings panel ala VSCode in the future.

ravicious commented on packages/teleterm/src/ui/appContext.ts:
```diff
@@ -147,4 +149,25 @@ export default class AppContext implements IAppContext {
+  private showConfigParsingErrors(): void {
+    const errors = this.mainProcessClient.configService.getStoredConfigErrors();
+    if (errors) {
+      errors.forEach(error => {
+        this.notificationsService.notifyError({
+          title: `Invalid config value for key ${error.path[0]}`,
+          description: error.message,
+        });
+      });
```
Somehow I managed to not submit a comment about that, even though I vividly remember writing it two days ago, sorry.

---

It's not really that important at the moment as we have only a single config option. But once we add more than a few options, it'd be good to send a single notification for all errors rather than a notification for each error.

Our notification system doesn't handle displaying many notifications that well.

For now I'm fine with leaving a TODO comment there and coming back to it later.

gzdunek replied:
Added single notification!

ravicious reviewed (approved):
I haven't really tested the implementation myself, but I believe if there's any rough edges we're not aware of, we'll have a chance to polish them once you add code which actually fully utilizes the new config system.

ravicious commented:
#1450 has to be backported after this PR.
